src/Symfony/Component/Process/Process.php

```diff
@@ -1632,6 +1635,17 @@ private function escapeArgument(?string $argument): string
         return '"'.str_replace(['"', '^', '%', '!', "\n"], ['""', '"^^"', '"^%"', '"^!"', '!LF!'], $argument).'"';
     }
 
+    private function replacePlaceholders(string $commandline, array $env)
+    {
+        return preg_replace_callback('/"\$([_a-zA-Z]++[_a-zA-Z0-9]*+)"/', function ($matches) use ($commandline, $env) {
+            if (!isset($env[$matches[1]]) || false === $env[$matches[1]]) {
+                throw new InvalidArgumentException(sprintf('Command line is missing a value for key %s: no values provided for any placeholders.', $matches[0]));
+            }
+
+            return '\\' === \DIRECTORY_SEPARATOR ? sprintf('!%s!', $matches[1]) : $matches[0];
```

Comment:
instead of a !FOO! reference, we should use escapeArgument

Reply:
The test breaks on windows :/

Reply:
The test was wrong, I fixed it.